### PR TITLE
Define Generic Permissions

### PR DESCRIPTION
**Pull request by ethanfrey**
I want to generalize the `Address` concept to a more powerful `Permission` than can capture more than just public key signatures, like hash locks, or approvals by other modules.

- [x] Spec in docs/design
- [x] Create Permission type
- [x] Migrate Authenticator to use Permission
- [x] Update sigs logic, stdsignature
- [x] Update cash
- [x] Fix any remaining tests

### CODE REVIEW

File: docs/design/permissions.rst [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+the "requester", which may be made of signatures, preimages,
+or other objects.
+
+*Authorization* happens in a handler, and after it decides that
+the are sufficient to execute this transaction. This is a
+comparison of the "requester" with the required permissions
+
+*Permissions* are a way to capture which conditions a transaction
+must fulfill to be able to execute an action. They must be
+serializable and can be stored along with an object.
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply docs/design/permissions.rst]: Yes, I think the word is a bit confusing. It is something similar, as I would use `request.HasPermission("sigs.ed25519.DEADBEEF....")`, but yeah, it is checking what power has been assigned to the request (by virtue of validating those bytes) and not representing the right to do something. Eg. the right to act as arbiter of an escrow is a permission in your model, and I refer to the precondition that enables access to that right as a permission.

This is a good point, and I would be happy if you could suggest a better term, or even better, a paragraph explaining these terms, including the one you suggest.

File: x/auth.go [hunk 2]
```diff
@@ -38,42 +40,68 @@ func (m MultiAuth) GetPermissions(ctx weave.Context) []weave.Address {
...
 	return false
 }
 
-// MainSigner returns the first signed if any, otherwise nil
-func MainSigner(ctx weave.Context, auth Authenticator) weave.Address {
+// GetAddresses wraps the GetPermissions method of any Authenticator
+func GetAddresses(ctx weave.Context, auth Authenticator) []weave.Address {
+	perms := auth.GetPermissions(ctx)
+	if len(perms) == 0 {
+		return nil
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/auth.go]: Good point, I don't think I really need nil...
I can check `len(addrs) == 0` as a generic guard. Not sure if there is any other point where `nil` is important to signal.

File: errors/common.go [hunk 3]
```diff
@@ -18,15 +18,16 @@ const (
...
-	errDecoding            = fmt.Errorf("Error decoding input")
-	errTooLarge            = fmt.Errorf("Input size too large")
-	errUnknownTxType       = fmt.Errorf("Tx type unknown")
-	errUnauthorized        = fmt.Errorf("Unauthorized")
-	errMissingSignature    = fmt.Errorf("Signature missing")
-	errInvalidSignature    = fmt.Errorf("Signature invalid")
-	errUnrecognizedAddress = fmt.Errorf("Unrecognized Address")
-	errInvalidChainID      = fmt.Errorf("Invalid ChainID")
-	errModifyChainID       = fmt.Errorf("Cannot modify ChainID")
+	errDecoding               = fmt.Errorf("Error decoding input")
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply errors/common.go]: I actually want these to be std errors, not pkg/errors, so they do not have a stack trace (if they did, the stack trace would be of the init function, as I have seen in some repos).

Trying to import `errors` and `github.com/pkg/errors` in the middle of my own `github.com/confio/weave/errors` package seemed like too much confusion. So I used `fmt.Errorf` to get rid of one of those imports.

How would you deal with this better?

File: docs/design/permissions.rst [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+The simplest example is "who can transfer money out of an account".
+In many blockchains, they hash the public key and use that
+to form an "address". Then this address is used as a primary key
+to an account balance. A user can send tokens to any address,
+and if I have signed with a public key, which hashes to the
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply docs/design/permissions.rst]: I guess I mean "sign with the private key corresponding to the public key that generated the address" but that was a mouthful. A better wording?

I can provide links to how ethereum and bitcoin addresses are formed:

https://en.bitcoin.it/wiki/Address#Address_map
https://kobl.one/blog/create-full-ethereum-keypair-and-address/#derive-the-ethereum-address-from-the-public-key

File: errors/common.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alpe) [reply errors/common.go]: ok, I can understand this. How about an import alias instead?
```
import (stderrors "errors" )
errDecoding = stderrors.New("Error decoding input")

File: docs/design/permissions.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (alpe) [reply docs/design/permissions.rst]: 👍 thanks for the links! the long description makes more sense for me. I stumbled upon the "public key signing" that didn't fit into my process understanding

File: docs/design/permissions.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (alpe) [reply docs/design/permissions.rst]: "Condition" as in the IETF draft may not have caused that wrong assumptions as "Permissions" does. when you put them into an auth package they would become `auth.Condition` and could be checked like `auth.MatchesAllConditions((ctx weave.Context, auth Authenticator, required []Condition) bool`. But let me think about it a bit more.

**Pull request merged**